cpp/src/arrow/compute/kernels/scalar_temporal_test.cc

```diff
@@ -2611,6 +2611,309 @@ TEST_F(ScalarTemporalTest, TestRoundTemporal) {
+    "2018-10-28 01:25:00", "2018-10-28 01:25:00", "2018-10-28 01:50:00",
+    "2018-10-28 01:50:00", "2018-10-28 01:50:00", "2018-10-28 02:15:00",
+    "2018-10-28 02:15:00", "2018-10-28 02:15:00", "2018-10-28 02:40:00"])";
+  const char* times_floor = R"([
+    "2018-10-27 22:20:00", "2018-10-27 22:45:00", "2018-10-27 22:45:00",
+    "2018-10-27 22:45:00", "2018-10-27 23:10:00", "2018-10-27 23:10:00",
+    "2018-10-27 23:10:00", "2018-10-27 23:35:00", "2018-10-27 23:35:00",
+    "2018-10-27 23:35:00", "2018-10-27 23:35:00", "2018-10-27 23:35:00",
+    "2018-10-27 23:35:00", "2018-10-27 23:45:00", "2018-10-27 23:45:00",
+    "2018-10-27 23:45:00", "2018-10-28 00:10:00", "2018-10-28 00:10:00",
+    "2018-10-28 00:10:00", "2018-10-28 00:35:00", "2018-10-28 00:35:00",
+    "2018-10-28 00:35:00", "2018-10-28 01:00:00", "2018-10-28 01:00:00",
+    "2018-10-28 01:00:00", "2018-10-28 01:25:00", "2018-10-28 01:25:00",
+    "2018-10-28 01:25:00", "2018-10-28 01:50:00", "2018-10-28 01:50:00",
+    "2018-10-28 01:50:00", "2018-10-28 02:15:00", "2018-10-28 02:15:00"])";
+  const char* times_round = R"([
+    "2018-10-27 22:45:00", "2018-10-27 22:45:00", "2018-10-27 22:45:00",
+    "2018-10-27 23:10:00", "2018-10-27 23:10:00", "2018-10-27 23:10:00",
+    "2018-10-27 23:35:00", "2018-10-27 23:35:00", "2018-10-27 23:35:00",
+    "2018-10-27 23:45:00", "2018-10-27 23:45:00", "2018-10-27 23:35:00",
+    "2018-10-27 23:35:00", "2018-10-28 00:10:00", "2018-10-28 00:10:00",
+    "2018-10-28 00:10:00", "2018-10-28 00:10:00", "2018-10-28 00:10:00",
+    "2018-10-28 00:35:00", "2018-10-28 00:35:00", "2018-10-28 00:35:00",
+    "2018-10-28 01:00:00", "2018-10-28 01:00:00", "2018-10-28 01:00:00",
+    "2018-10-28 01:25:00", "2018-10-28 01:25:00", "2018-10-28 01:25:00",
+    "2018-10-28 01:50:00", "2018-10-28 01:50:00", "2018-10-28 01:50:00",
+    "2018-10-28 02:15:00", "2018-10-28 02:15:00", "2018-10-28 02:15:00"])";
+
+  CheckScalarUnary("floor_temporal", unit, complete_times, unit, complete_times_floor,
+                   &options);
+  CheckScalarUnary("ceil_temporal", unit, times, unit, times_ceil, &options);
+  CheckScalarUnary("floor_temporal", unit, times, unit, times_floor, &options);
+  CheckScalarUnary("round_temporal", unit, times, unit, times_round, &options);
+}
+
+TEST_F(ScalarTemporalTest, TestCeilFloorRoundTemporalNonexistent1) {
+  // Asia/Tehran switches from UTC+3:30 to UTC+4:30 on 2022-03-22 00:00:00 UTC+3:30
+  // This causes an hour long non-existing period in local time.
+  auto unit = timestamp(TimeUnit::SECOND, "Asia/Tehran");
+  auto options = RoundTemporalOptions(16, CalendarUnit::MINUTE);
+  const char* times = R"([
+    "2022-03-21 19:30:00", "2022-03-21 20:00:00", "2022-03-21 20:30:00"])";
+  const char* times_ceil = R"([
+    "2022-03-21 19:42:00", "2022-03-21 20:14:00", "2022-03-21 20:34:00"])";
+  const char* times_floor = R"([
+    "2022-03-21 19:26:00", "2022-03-21 19:58:00", "2022-03-21 20:18:00"])";
```

Comment:
This still doesn't work: there are 20 minutes of local time between `2022-03-21 19:58:00` and `2022-03-21 20:18:00`, which is not a multiple of 16, so the two timestamps cannot be _both_ multiples of 16.